File: lib/effs/tls_close.c
// TLS close: orderly shutdown, session free and fd close. Best
// effort: always answers Unit. The handle is a TlsConn*.
#include <dlfcn.h>
#include <unistd.h>

#ifndef TLS_CONN_DEFINED
#define TLS_CONN_DEFINED
struct TlsConn {
  void* ssl;
  void* ctx;
  int   fd;
};
#endif

Term tls_close_run(Env e, Term* f, IoWork* w) {
  (void)e;
  (void)w;
  struct TlsConn* c = (struct TlsConn*)(intptr_t)io_hand_v(f[0]);
  if (c != NULL) {
    void* h = dlopen("libssl.so.3", RTLD_NOW);
    if (h == NULL) {
      h = dlopen("libssl.so", RTLD_NOW);
    }
    if (h != NULL) {
      int (*p_shutdown)(void*) = dlsym(h, "SSL_shutdown");
      void (*p_free)(void*) = dlsym(h, "SSL_free");
      void (*p_ctx_free)(void*) = dlsym(h, "SSL_CTX_free");
      if (p_shutdown != NULL && c->ssl != NULL) {
        p_shutdown(c->ssl);
      }
      if (p_free != NULL && c->ssl != NULL) {
        p_free(c->ssl);
      }
      if (p_ctx_free != NULL && c->ctx != NULL) {
        p_ctx_free(c->ctx);
      }
      dlclose(h);
    }
    if (c->fd >= 0) {
      close(c->fd);
    }
    free(c);
  }
  return term_pak(CID_UNIT, 0);
}

static void __attribute__((constructor)) tls_close_use(void) {
  // Unit answer, no handle back: plain effect, not a Result pair.
  io_eff(CID_TLS_CLOSE, tls_close_run, 0);
}

File: lib/effs/tls_close.js
// TLS close: orderly shutdown, session free and fd close. Best effort,
// always answers Unit. The handle is a TlsConn box.
function tls_close(socket) {
  const conn = socket;
  if (conn && conn.$ === "TlsConn") {
    const lib = tls_lib();
    if (lib !== null) {
      try {
        lib.SSL_shutdown(conn.ssl);
      } catch (e) {}
      try {
        lib.SSL_free(conn.ssl);
      } catch (e) {}
      try {
        lib.SSL_CTX_free(conn.ctx);
      } catch (e) {}
    }
    try {
      io_sys().close(conn.fd);
    } catch (e) {}
    try {
      if (conn._cb) conn._cb.close();
    } catch (e) {}
  }
  return { $: "Unit" };
}

// The libssl symbols this effect needs, dlopened once per process and
// cached in a global slot (the dns_file style).
function tls_lib() {
  if (globalThis.BEND_TLS_CLOSE === undefined) {
    try {
      const ffi = require("bun:ffi");
      const T = { i: "i32", p: "ptr", v: "void" };
      const lib = ffi.dlopen("libssl.so.3", {
        SSL_shutdown: { args: [T.p], returns: T.i },
        SSL_free: { args: [T.p], returns: T.v },
        SSL_CTX_free: { args: [T.p], returns: T.v },
      });
      globalThis.BEND_TLS_CLOSE = { ...lib.symbols };
    } catch (e) {
      globalThis.BEND_TLS_CLOSE = null;
    }
  }
  return globalThis.BEND_TLS_CLOSE;
}